### PR TITLE
MAINT: stats: `nan_policy` for multidimensional non-NumPy arrays

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-24811
Toward gh-14651

#### What does this implement/fix?
Previously, most stats functions ignored `nan_policy` with multidimensional non-NumPy arrays and behaved consistently with `nan_policy='propagate'`. This PR:
- documents that `nan_policy='raise'` is not supported by lazy backends;
- documents that for multidimensional input, `nan_policy='omit'` is compatible only with the NumPy backend, suggesting MArray as an alternative; and
- ensures that `nan_policy='raise'` and `nan_policy='propagate'` are respected for multidimensional input.

#### Additional information
The goal is to run all backends through something like `test_axis_nan_policy.py`, but that is left to future work (tracked by gh-14651).

#### AI Generation Disclosure
No AI

### CODE REVIEW

File: scipy/stats/tests/test_stats.py [hunk 1]
```diff
@@ -5619,10 +5619,14 @@ def test_ttest_many_dims(self, kwds, equal_var):
         assert_allclose(statistics, res.statistic)
         assert_allclose(pvalues, res.pvalue)
 
+    @make_xp_test_case(stats.ttest_ind)
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:5825]: At some point, should revisit array API support and testing in `ttest_ind` - add support for `trim`, collect all tests into a class, etc.

File: scipy/stats/_axis_nan_policy.py [hunk 2]
```diff
@@ -616,14 +620,24 @@ def hypotest_fun_out(*samples, **kwds):
                 res = _add_reduced_axes(res, reduced_axes, keepdims)
                 return tuple_to_result(*res)
 
-            if not is_numpy(xp) and 'nan_policy' in kwds:
-                msg = ("Use of `nan_policy` is incompatible with multidimensional "
-                       "non-NumPy arrays.")
```
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:625]: This wasn't working before because `'nan_policy'` has been `pop`ped out of `kwds`.

File: scipy/stats/_axis_nan_policy.py [hunk 3]
```diff
@@ -616,14 +620,24 @@ def hypotest_fun_out(*samples, **kwds):
...
+                msg = ("Use of `nan_policy='omit'` is incompatible with "
+                       "multidimensional non-NumPy arrays. Similar behavior may be "
+                       "supported by masking NaNs with MArray; see Notes.")
                 raise NotImplementedError(msg)
 
             if not is_numpy(xp):
-                res = hypotest_fun_out(*samples, axis=axis, **kwds)
+                # At this point, we know axis=-1 unconditionally
+                x = _broadcast_concatenate(samples, axis=-1, paired=paired)
+                contains_nan = _contains_nan(x, nan_policy)
```
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:637]: This handles `nan_policy='raise'`.

File: scipy/stats/_axis_nan_policy.py [hunk 4]
```diff
@@ -616,14 +620,24 @@ def hypotest_fun_out(*samples, **kwds):
...
+                res = hypotest_fun_out(*samples, axis=-1, **kwds)
                 res = result_to_tuple(res, n_out)
+
+                if is_lazy_array(x) or contains_nan:
+                    nan_out = xp.any(xp.isnan(x), axis=-1)
+                    res = ((xpx.at(res_i)[nan_out].set(xp.nan, copy=True)
+                            if hasattr(res_i, 'shape') else res_i) for res_i in res)
```
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py]: This handles `nan_policy='propagate'`. We check for `'shape'` because some attributes of the result object are not arrays.

File: scipy/stats/_axis_nan_policy.py [hunk 5]
```diff
@@ -616,14 +620,24 @@ def hypotest_fun_out(*samples, **kwds):
...
 
             if not is_numpy(xp):
-                res = hypotest_fun_out(*samples, axis=axis, **kwds)
+                # At this point, we know axis=-1 unconditionally
+                x = _broadcast_concatenate(samples, axis=-1, paired=paired)
+                contains_nan = _contains_nan(x, nan_policy)
+                res = hypotest_fun_out(*samples, axis=-1, **kwds)
                 res = result_to_tuple(res, n_out)
+
+                if is_lazy_array(x) or contains_nan:
```
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py]:
```suggestion
                if override['nan_propagation'] and (is_lazy_array(x) or contains_nan):
```

**Review by mdhaber — COMMENTED** [review]
Some self-review.

File: scipy/stats/_axis_nan_policy.py [hunk 6]
```diff
@@ -316,8 +316,12 @@ def _get_axis_params(default_axis=0, _name=_name, _desc=_desc):  # bind NOW
...
   statistic is computed, the corresponding entry of the output will be
   NaN.
-- ``raise``: if a NaN is present, a ``ValueError`` will be raised."""
-    .split('\n'))
+- ``raise``: if a NaN is present, a ``ValueError`` will be raised.
+
+``nan_policy='raise'`` is not supported by lazy backends. For multidimensional input,
+``nan_policy='omit'`` is supported only by the NumPy backend. An alternative is to
+mask NaN values using `MArray <https://mdhaber.github.io/marray/tutorial.html>`__; see
+Notes for support information.""".split('\n'))
```
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py]:
```suggestion
``nan_policy='raise'`` and ``nan_policy='omit'`` are not supported by lazy backends, and
``nan_policy='omit'`` is not compatible with MArray input. For multidimensional input,
``nan_policy='omit'`` is supported only by the NumPy backend. Instead,
mask NaN values using `MArray <https://mdhaber.github.io/marray/tutorial.html>`__;
and use the default `nan_policy`. See Notes for support information.""".split('\n'))
```

File: scipy/stats/tests/test_stats.py (hunk 1 above)
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:5825]: Done.

**Comment by j-bowhay** [comment]
Failure looks related

**Comment by mdhaber** [comment]
Yeah, probably messed something up during merge.

**Comment by atharvadarke** [comment]
Hi! It looks like this issue is already being addressed by PR #24813. Can this issue be closed?

**Comment by mdhaber** [comment]
Oops, probably didn't need to merge again; merge two days ago showed no failures. This is pretty important for 2.0, I think.

File: scipy/stats/_axis_nan_policy.py [hunk 7]
```diff
@@ -620,14 +625,24 @@ def hypotest_fun_out(*samples, **kwds):
...
+                # At this point, we know axis=-1 unconditionally
+                x = _broadcast_concatenate(samples, axis=-1, paired=paired)
+                contains_nan = _contains_nan(x, nan_policy)
+                res = hypotest_fun_out(*samples, axis=-1, **kwds)
                 res = result_to_tuple(res, n_out)
+
+                if override['nan_propagation'] and (is_lazy_array(x) or contains_nan):
+                    nan_out = xp.any(xp.isnan(x), axis=-1)
+                    res = ((xp.where(nan_out, xp.nan, res_i) if hasattr(res_i, 'shape')
+                            else res_i) for res_i in res)
```
Comment (lucascolley) [inline scipy/stats/_axis_nan_policy.py:650]: could you explain what this means for scalar `res_i`?

File: scipy/stats/tests/test_stats.py [hunk 8]
```diff
@@ -6358,13 +6379,19 @@ def test_ttest_1samp_new(self, xp):
...
-            xp_assert_equal(res.statistic, xp.asarray([0., xp.nan]))
-            xp_assert_equal(res.pvalue, xp.asarray([1., xp.nan]))
+            if not is_array_api_strict(xp):
+                # The `_axis_nan_policy` decorator wants to propagate NaNs into integer
+                # degrees of freedom, but array-api-strict refuses to promote between
+                # integers and floats. This is part of a larger dtype instability issue
+                # for integer outputs of stats functions; skip for now.
 
-    @skip_xp_backends(eager_only=True, reason="lazy -> reduced nan_policy capabilities")
+                # check that nan in input array result in nan output
```
Comment (lucascolley) [inline scipy/stats/tests/test_stats.py]:
```suggestion
                # check that nan in input array results in nan output
```

Comment (mdhaber) [reply scipy/stats/tests/test_stats.py]: ~~Fixed locally. Will commit soon.~~ Done.

File: scipy/stats/_axis_nan_policy.py (hunk 7 above)
Comment (mdhaber) [reply scipy/stats/_axis_nan_policy.py:650]: ~~I'm investigating this. I think I remember but am checking and will get back to you soon.~~
I added a comment explains that the t-test functions stores their `alternative` in the result object as a Python scalar and we don't want NaNs to "propagate" into it. That is sort of a hack we added a while back (gh-16835 I think) to work around limitations of the `_axis_nan_policy` decorator, so we work around that hack here. Nothing else minds it being just `(xp.where(nan_out, xp.nan, res_i) for res_i in res)`.

In the meantime, if it is convenient to run `scipy/stats/tests/test_morestats::TestWilcoxon` locally with JAX backend, do you get any failures (in `main` or this branch)?

Comment (j-bowhay) [reply scipy/stats/_axis_nan_policy.py:650]: Both pass for me

Comment (mdhaber) [reply scipy/stats/_axis_nan_policy.py:650]: Huh. Well that's not related to this PR, so I'll open a separate issue about it. What version of JAX do you have, Jake?

File: scipy/stats/_axis_nan_policy.py [hunk 9]
```diff
@@ -620,14 +625,30 @@ def hypotest_fun_out(*samples, **kwds):
...
+                res = hypotest_fun_out(*samples, axis=-1, **kwds)
                 res = result_to_tuple(res, n_out)
+
+                if override['nan_propagation'] and (is_lazy_array(x) or contains_nan):
+                    nan_out = xp.any(xp.isnan(x), axis=-1)
+                    # this could be
+                    # (xp.where(nan_out, xp.nan, res_i) for res_i in res)
+                    # but `ttest_ind` stores its `alternative`` as a scalar integer
+                    # in the test result. That is sort of a hack and could be improved
+                    # at some point, but in the meantime, leave "all" Python scalars -
```
Comment (lucascolley) [inline scipy/stats/_axis_nan_policy.py:647]: can we open an issue for this?

Comment (mdhaber) [reply scipy/stats/_axis_nan_policy.py:647]: I'd rather not? 
I don't think we'd want to put more time into polishing the existing `_axis_nan_policy` decorator. After years of working with `nan_policy`, I don't think `nan_policy='omit'` should exist. There should be a distinction between NaNs and missing values, and that is provided by masked arrays. With `_apply_over_axis` and without `nan_policy='omit'`, the `_axis_nan_policy` decorator could be rewritten pretty compactly and avoid this hack. In the meantime, this is not bothering me, anyway.

Comment (lucascolley) [reply scipy/stats/_axis_nan_policy.py:647]: alright

File: scipy/stats/_axis_nan_policy.py (hunk 7 above)
Comment (j-bowhay) [reply scipy/stats/_axis_nan_policy.py:650]:
```
In [1]: import jax

In [2]: jax.__version__
Out[2]: '0.7.2'
```

Comment (mdhaber) [reply scipy/stats/_axis_nan_policy.py:650]: Oh, wow : ) ok, it probably broke with JAX 0.10. I wonder if our CI is still using an older version. Anyway, I'll look into that in a separate issue since it's failing in main, too.

Comment (j-bowhay) [reply scipy/stats/_axis_nan_policy.py:650]: Sounds plausible, I am just using whatever version `pixi` gives me

**Review by lucascolley — APPROVED** [review]
thanks Matt, happy with this now

**Comment by mdhaber** [comment]
Looks like this is causing the `test_power_divergence_chisquare` failure when there is an all-masked slice. CI didn't fail here because that test was added recently. I'm taking a look.

File: scipy/stats/_axis_nan_policy.py [hunk 10]
```diff
@@ -620,14 +625,30 @@ def hypotest_fun_out(*samples, **kwds):
...
+                    # which is just the `alternative` of `ttest_ind` AFAICT - alone.
+                    res = ((xp.where(nan_out, xp.nan, res_i) if hasattr(res_i, 'shape')
+                            else res_i) for res_i in res)
```
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:650]: This can replace a masked value with an unmasked NaN, hence the failure. Probably needs to preserve the mask.
